adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenByDeviceCodeHandler.cs

```diff
@@ -49,10 +50,13 @@ public AcquireTokenByDeviceCodeHandler(RequestData requestData, DeviceCodeResult
         {
             TimeSpan timeRemaining = deviceCodeResult.ExpiresOn - DateTimeOffset.UtcNow;
             AdalResultWrapper resultEx = null;
+
             //the interval is added so that the while loop does not end before aquiring the last response from the STS stating that the code has expired.
             //Without it, resultEx will end up being null.
-            while (timeRemaining.TotalSeconds + deviceCodeResult.Interval > 0)
```

Comment:
This condition can fail, probably not in real scenarios where the timeouts are large, but it did fail in tests. The underlying issue is that we should contact the server at least once. Alternatively if we notice the timeout has already happened, we could timeout directly, but this requires more code and behaviour change.

Reply:
I have a few concerns with this approach. First, I would consider sending a request to service with expired code as a Service contract violation. Second, why would we try to use expired code when we already know it is expired? Also I see the idea of changing business logic because some unit test is failing as very dangerous. From my perspective, you can just increase timeout in test data - it is a unit test - we can mock any server response with any parameters. I think this would be a solution.

Also, why do you suggest we contact the server at least once? I would like to know more about your reasoning here.

Reply:
I would also agree with @SomkaPe that ideally we should not need to send out that request when we already know the code has expired. And, to be fair, @bgavrilMS did mention "alternatively if we notice the timeout has already happened, we could timeout directly", so he also saw this, hence not necessarily "suggest" us to "contact server at least once".

The underlying issue, imo, is not that we *should* contact the server at least once. It is that our pre-existing code happened to [already be written to expect a result available](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1123/files#diff-c04b555b31a145194bbca57706839214R205) and then we kinda had to [follow it](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1123/files#diff-ac38c278be1bc474043033ed72031a67R54).

In terms of the solution/workaround this time, given that this condition rarely happens in real life, can we actually consider simply raise an exception for the timeout, and catch it in the test case, and then call it a day?

Reply:
+1 for the JDD reference :)